Core/GCUniqueCommitCollection.h

```diff
@@ -0,0 +1,43 @@
+//  Copyright (C) 2015 Pierre-Olivier Latour <info@pol-online.net>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#import <Foundation/Foundation.h>
+
+@class GCCommit;
+
+
+/**
+ * This class represents a collection of commits with preserved order.
+ * It is also optimized to be fast when you have a lot of calls to
+ * - addCommit: , -hasCommit: and -removeCommit: methods.
+ */
+@interface GCUniqueCommitCollection : NSObject
```

Comment:
Current implementation is a quite generic one, I agree.
The difference between ordered set is that when you re-add a commit it preserves the old position. It doesn't really matter in our case (we do not re-add commits, if memory serves), but it might cause some issues in future. I don't know if it is worth making too generic... But it might be a good point, I will think about it.

It also uses SHA1 method,  so it probably should accept `GCObject*` then, not `id`.